core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -946,17 +937,34 @@ public int getSlaveAgentPort() {
      */
     public void setSlaveAgentPort(int port) throws IOException {
         this.slaveAgentPort = port;
+        launchTcpSlaveAgentListener();
+    }
 
-        // relaunch the agent
-        if(tcpSlaveAgentListener==null) {
-            if(slaveAgentPort!=-1)
+    private void launchTcpSlaveAgentListener() throws IOException {
+        // shutdown previous agent if the port has changed
+        if(tcpSlaveAgentListener != null && tcpSlaveAgentListener.configuredPort!=slaveAgentPort) {
+            tcpSlaveAgentListener.shutdown();
+            tcpSlaveAgentListener = null;
+        }
+        if(slaveAgentPort!=-1 && tcpSlaveAgentListener == null) {
+            String administrativeMonitorId = getClass().getName()+".tcpBind";
+            try {
                 tcpSlaveAgentListener = new TcpSlaveAgentListener(slaveAgentPort);
-        } else {
-            if(tcpSlaveAgentListener.configuredPort!=slaveAgentPort) {
-                tcpSlaveAgentListener.shutdown();
-                tcpSlaveAgentListener = null;
-                if(slaveAgentPort!=-1)
-                    tcpSlaveAgentListener = new TcpSlaveAgentListener(slaveAgentPort);
+                // remove previous monitor in case of previous error
+                AdministrativeMonitor previousMonitor = null;
+                for(AdministrativeMonitor am : AdministrativeMonitor.all()) {
+                    if (administrativeMonitorId.equals(am.id)) {
+                        previousMonitor = am;
+                        break;
+                    }
+                }
+                if (previousMonitor != null) {
+                    AdministrativeMonitor.all().remove(previousMonitor);
```

Comment:
Also, if `new TcpSlaveAgentListener(slaveAgentPort)` throws `BindException`, the old error will not be removed which is I dare to say undesirable.  How about moving the cleaning logic into `AdministrativeError` registration so it will consider new errors overwrite old ones in case of ID conflict?